README.md
[![Dart](https://github.com/devoncarew/vtable/actions/workflows/ci.yaml/badge.svg)](https://github.com/devoncarew/vtable/actions/workflows/ci.yaml)
[![pub package](https://img.shields.io/pub/v/vtable.svg)](https://pub.dev/packages/vtable)
[![package publisher](https://img.shields.io/pub/publisher/vtable.svg)](https://pub.dev/packages/vtable/publisher)

A Flutter table widget featuring virtualization, sorting, and custom cell
rendering.

## Features

A Flutter table widget featuring virtualization, sorting, and custom cell
rendering. Features include:

- virtualization
- sorting
- custom cell rendering
- cell tooltips
- customized column widths and grow
- cell alignment
- cell validation
- custom widgets for various table areas (table summary, filtering widgets,
  action widgets)

![Screenshot](doc/screenshot.png)

## Getting started

From the command line:

    flutter pub add vtable

And in your project source:

    import 'package:vtable/vtable.dart';

## Usage

```dart
Widget build(BuildContext context) {
  return VTable<SampleRowData>(
    items: listOfItems,
    columns: [
      VTableColumn(
        label: 'Planet',
        width: 120,
        grow: 0.6,
        transformFunction: (row) => row.name,
      ),
      VTableColumn(
        label: 'Gravity',
        width: 100,
        grow: 0.3,
        transformFunction: (row) => row.gravity.toStringAsFixed(1),
        alignment: Alignment.centerRight,
        compareFunction: (a, b) => a.gravity.compareTo(b.gravity),
        validators: [SampleRowData.validateGravity],
      ),
    ],
  );
}
```

See also the
[main.dart](https://github.com/devoncarew/vtable/blob/main/example/main.dart)
example and the package's [API](https://pub.dev/documentation/vtable/latest/)
documentation.


CHANGELOG.md
## 0.4.0

 * Add a `showHeaders` parameter to `VTable`.
 * Add a `showToolbar` parameter to `VTable`.
 * Improve the display of the table in dark themes.

## 0.3.1

* Use the `Theme.dividerColor` color to separate rows.

## 0.3.0

* Make the row height configurable; add a `VTable.rowHeight` parameter.

## 0.2.0

* Add a `includeCopyToClipboardAction` flag.
* Add dartdocs.

## 0.1.1

* Update markdown badges for the first publish.

## 0.1.0

* Initial release.


example/main.dart
// Copyright (c) 2023, Devon Carew. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:vtable/vtable.dart';

void main() {
  runApp(
    ExampleApp(
      items: generateRowData(planets.length * 10),
    ),
  );
}

class ExampleApp extends StatefulWidget {
  final List<SampleRowData> items;

  const ExampleApp({
    required this.items,
    super.key,
  });

  @override
  State<ExampleApp> createState() => _ExampleAppState();
}

class _ExampleAppState extends State<ExampleApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'VTable Example App',
      home: Scaffold(
        appBar: AppBar(
          title: const Text('VTable Example App'),
        ),
        body: Center(
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 8.0),
            child: createTable(),
          ),
        ),
      ),
      debugShowCheckedModeBanner: false,
    );
  }

  VTable<SampleRowData> createTable() {
    const disabledStyle =
        TextStyle(fontStyle: FontStyle.italic, color: Colors.grey);

    return VTable<SampleRowData>(
      items: widget.items,
      tableDescription: '${widget.items.length} items',
      startsSorted: true,
      includeCopyToClipboardAction: true,
      columns: [
        VTableColumn(
          label: 'ID',
          width: 180,
          grow: 1,
          transformFunction: (row) => row.id,
        ),
        VTableColumn(
          label: 'Planet',
          width: 100,
          grow: 1,
          transformFunction: (row) => row.planet.name,
          styleFunction: (row) => row.planet == moon ? disabledStyle : null,
        ),
        VTableColumn(
          label: 'Gravity (m/s²)',
          width: 120,
          grow: 1,
          transformFunction: (row) => row.planet.gravity.toStringAsFixed(1),
          alignment: Alignment.centerRight,
          compareFunction: (a, b) =>
              a.planet.gravity.compareTo(b.planet.gravity),
          validators: [SampleRowData.validateGravity],
        ),
        VTableColumn(
          label: 'Orbit distance (AU)',
          width: 120,
          grow: 1,
          transformFunction: (row) =>
              (row.planet.orbit / earth.orbit).toStringAsFixed(1),
          alignment: Alignment.centerRight,
          compareFunction: (a, b) => a.planet.orbit.compareTo(b.planet.orbit),
        ),
        VTableColumn(
          label: 'Orbital period (years)',
          width: 140,
          grow: 1,
          transformFunction: (row) =>
              (row.planet.period / earth.period).toStringAsFixed(1),
          alignment: Alignment.centerRight,
          compareFunction: (a, b) => a.planet.period.compareTo(b.planet.period),
        ),
        VTableColumn(
          label: 'Moons',
          width: 100,
          grow: 1,
          transformFunction: (row) => row.planet.moons.toString(),
          alignment: Alignment.centerRight,
          compareFunction: (a, b) => a.planet.moons - b.planet.moons,
        ),
        VTableColumn(
          label: 'Temperature (C)',
          width: 120,
          transformFunction: (row) => row.planet.temp.toString(),
          alignment: Alignment.centerRight,
          compareFunction: (a, b) => a.planet.temp - b.planet.temp,
        ),
        VTableColumn(
          label: 'Temperature',
          width: 120,
          alignment: Alignment.center,
          transformFunction: (row) => row.planet.temp.toString(),
          compareFunction: (a, b) => a.planet.temp - b.planet.temp,
          renderFunction: (context, data, _) {
            Color color;
            if (data.planet.temp < 0) {
              color = Colors.blue
                  .withAlpha((data.planet.temp / Planet.coldest * 255).round());
            } else {
              color = Colors.red
                  .withAlpha((data.planet.temp / Planet.hotest * 255).round());
            }
            return Chip(
              label: const SizedBox(width: 48),
              backgroundColor: color,
            );
          },
        ),
      ],
    );
  }
}

List<SampleRowData> generateRowData(int rows) {
  final words = loremIpsum
      .toLowerCase()
      .replaceAll(',', '')
      .replaceAll('.', '')
      .split(' ');
  final random = math.Random();

  return List.generate(rows, (index) {
    final word1 = words[random.nextInt(words.length)];
    final word2 = words[random.nextInt(words.length)];
    final val = random.nextInt(10000);
    final id = '$word1-$word2-${val.toString().padLeft(4, '0')}';

    return SampleRowData(
      id: id,
      planet: planets[random.nextInt(planets.length)],
    );
  });
}

const String loremIpsum =
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod '
    'tempor incididunt ut labore et dolore magna aliqua.';

const Planet earth = Planet('Earth', 9.8, 149.6, 365.2, 15, 1);
const Planet moon = Planet('Moon', 1.6, 0.384, 27.3, -20, 0);

const List<Planet> planets = <Planet>[
  Planet('Mercury', 3.7, 57.9, 88, 167, 0),
  Planet('Venus', 8.9, 108.2, 224.7, 464, 0),
  earth,
  moon,
  Planet('Mars', 3.7, 228, 687, -65, 2),
  Planet('Jupiter', 23.1, 778.5, 4331, -110, 92),
  Planet('Saturn', 9, 1432, 10747, -140, 83),
  Planet('Uranus', 8.7, 2867, 30589, -195, 27),
  Planet('Neptune', 11, 4515, 59800, -200, 14),
  Planet('Pluto', 0.7, 5906.4, 90560, -225, 5),
];

class Planet {
  final String name;
  final double gravity;
  final double orbit;
  final double period;
  final int temp;
  final int moons;

  const Planet(
    this.name,
    this.gravity,
    this.orbit,
    this.period,
    this.temp,
    this.moons,
  );

  static int get coldest =>
      planets.fold(0, (previous, next) => math.min(previous, next.temp));

  static int get hotest =>
      planets.fold(0, (previous, next) => math.max(previous, next.temp));

  @override
  String toString() => name;
}

class SampleRowData {
  final String id;
  final Planet planet;

  SampleRowData({required this.id, required this.planet});

  static ValidationResult? validateGravity(SampleRowData row) {
    if (row.planet.gravity > 20.0) {
      return ValidationResult.error('too heavy!');
    }
    if (row.planet.gravity > 10.0) {
      return ValidationResult.warning('pretty heavy');
    }
    return null;
  }

  @override
  String toString() => '$id (${planet.name})';
}


lib/src/copy_action.dart
// Copyright (c) 2023, Devon Carew. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../vtable.dart';
import 'theme.dart';

/// Copies the contents of a [VTable] into the clipboard.
///
/// This widget - when added to the `actions` of a [VTable] - will copy the
/// contents of the table into the clipboard in CSV format.
///
/// [T] is the row type of the target table.
///
/// Note that the parameterized type of the action must be provided in order
/// for the action to locate the table in the widget tree (so, for a
/// `VTable<MyRowType> table, the action should be constructed as
/// `CopyToClipboardAction<MyRowType>`).
class CopyToClipboardAction<T> extends StatelessWidget {
  const CopyToClipboardAction({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return IconButton(
      icon: const Icon(Icons.copy),
      tooltip: 'Copy table data to clipboard',
      iconSize: defaultIconSize,
      splashRadius: defaultSplashRadius,
      onPressed: () => _copyTableToClipboard(context),
    );
  }

  void _copyTableToClipboard(BuildContext context) {
    final table = context.findAncestorWidgetOfExactType<VTable<T>>()!;

    final buf = StringBuffer();

    // write out the column titles
    buf.writeln(table.columns.map((c) => c.label).join(','));

    // write out each row
    for (final item in table.items) {
      buf.writeln(table.columns.map((column) {
        String val = column.transformFunction != null
            ? column.transformFunction!(item)
            : '$item';
        return val.contains(',') ? '"$val"' : val;
      }).join(','));
    }

    Clipboard.setData(ClipboardData(text: buf.toString()));

    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Copied as CSV to clipboard.')),
    );
  }
}


lib/vtable.dart
// Copyright (c) 2023, Devon Carew. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:vtable/src/copy_action.dart';

import 'src/theme.dart';

/// A callback to react to row taps.
typedef ItemTapHandler<T> = void Function(T object);

/// A callback to react to table row selection changes.
typedef OnSelectionChanged<T> = void Function(T? selectedItem);

/// A Flutter table widget featuring virtualization, sorting, and custom cell
/// rendering.
class VTable<T> extends StatefulWidget {
  static const double _vertPadding = 4;
  static const double _horizPadding = 8;

  /// The list of data rows for the table.
  final List<T> items;

  /// The list of columns to display for the table.
  final List<VTableColumn<T>> columns;

  /// Whether the table should initially sort the rows.
  final bool startsSorted;

  /// Whether this table should track and render row selection.
  final bool supportsSelection;

  /// A callback used to react to a double tap on a table row.
  final ItemTapHandler<T>? onDoubleTap;

  /// A callback used to react to table selection changes.
  final OnSelectionChanged<T>? onSelectionChanged;

  /// A optional description of the table.
  ///
  /// If provided, this is displayed in the far left side of the table's
  /// toolbar. E.g., `'234 items'`.
  final String? tableDescription;

  /// An optional set of widgets to display in the left side of the table's
  /// toolbar.
  ///
  /// This is often used as a place to put filter related widgets.
  final List<Widget> filterWidgets;

  /// An optional set of widgets to display in the right side of the table's
  /// toolbar.
  ///
  /// This is often used as a place to display additional table actions.
  final List<Widget> actions;

  /// The delay to use for table tooltips.
  ///
  /// Tooltips are used to display any validation results from a column's
  /// validators.
  final Duration tooltipDelay;

  /// The height to use for table rows.
  final double rowHeight;

  /// Whether to include a 'copy to clipboard' action in the toolbar.
  ///
  /// If included, this action will copy the table's contents as CSV to the
  /// system clipboard.
  final bool includeCopyToClipboardAction;

  /// Whether to show column headers.
  final bool showHeaders;

  /// Whether to show a table toolbar.
  final bool showToolbar;

  /// Construct a new `VTable` instance.
  ///
  /// E.g:
  ///
  /// ```
  /// Widget build(BuildContext context) {
  ///   return VTable<SampleRowData>(
  ///     items: listOfItems,
  ///     columns: [
  ///       VTableColumn(
  ///         label: 'Planet',
  ///         width: 120,
  ///         grow: 0.6,
  ///         transformFunction: (row) => row.name,
  ///       ),
  ///       VTableColumn(
  ///         label: 'Gravity',
  ///         width: 100,
  ///         grow: 0.3,
  ///         transformFunction: (row) => row.gravity.toStringAsFixed(1),
  ///         alignment: Alignment.centerRight,
  ///         compareFunction: (a, b) => a.gravity.compareTo(b.gravity),
  ///         validators: [SampleRowData.validateGravity],
  ///       ),
  ///     ],
  ///   );
  /// }
  /// ```
  const VTable({
    required this.items,
    required this.columns,
    this.startsSorted = false,
    this.supportsSelection = false,
    this.onDoubleTap,
    this.onSelectionChanged,
    this.tableDescription,
    this.filterWidgets = const [],
    this.actions = const [],
    this.tooltipDelay = defaultTooltipDelay,
    this.rowHeight = defaultRowHeight,
    this.includeCopyToClipboardAction = false,
    this.showHeaders = true,
    this.showToolbar = true,
    Key? key,
  }) : super(key: key);

  @override
  State<VTable> createState() => _VTableState<T>();
}

class _VTableState<T> extends State<VTable<T>> {
  late ScrollController scrollController;
  late List<T> sortedItems;
  int? sortColumnIndex;
  bool sortAscending = true;
  final ValueNotifier<T?> selectedItem = ValueNotifier(null);

  @override
  void initState() {
    super.initState();

    scrollController = ScrollController();
    sortedItems = widget.items.toList();

    _performInitialSort();

    selectedItem.addListener(() {
      if (widget.onSelectionChanged != null) {
        widget.onSelectionChanged!(selectedItem.value);
      }
    });
  }

  @override
  void didUpdateWidget(VTable<T> oldWidget) {
    super.didUpdateWidget(oldWidget);

    sortedItems = widget.items;
    _performInitialSort();

    // Clear the selection if the selected item is no longer in the table.
    if (selectedItem.value != null &&
        !sortedItems.contains(selectedItem.value)) {
      selectedItem.value = null;
    }
  }

  void _performInitialSort() {
    if (widget.startsSorted && columns.first.supportsSort) {
      columns.first.sort(sortedItems, ascending: true);
      sortColumnIndex = 0;
    }
  }

  List<VTableColumn<T>> get columns => widget.columns;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        if (widget.showToolbar) createActionRow(context),
        if (widget.showToolbar) const Divider(),
        Expanded(
          child: LayoutBuilder(
            builder: (BuildContext context, BoxConstraints constraints) {
              Map<VTableColumn, double> colWidths = _layoutColumns(constraints);

              return Column(
                children: [
                  if (widget.showHeaders) createHeaderRow(colWidths),
                  Expanded(
                    child: createRowsListView(context, colWidths),
                  ),
                ],
              );
            },
          ),
        ),
      ],
    );
  }

  Padding createActionRow(BuildContext context) {
    final extraActions = [];

    if (widget.includeCopyToClipboardAction) {
      if (widget.actions.isNotEmpty) {
        extraActions.add(
            const SizedBox(height: toolbarHeight, child: VerticalDivider()));
      }

      extraActions.add(CopyToClipboardAction<T>());
    }

    return Padding(
      padding: const EdgeInsets.only(left: 8, top: 16),
      child: Row(
        children: [
          Text(widget.tableDescription ?? ''),
          const Expanded(child: SizedBox(width: 8)),
          ...widget.filterWidgets,
          const Expanded(child: SizedBox(width: 8)),
          ...widget.actions,
          ...extraActions,
        ],
      ),
    );
  }

  Row createHeaderRow(Map<VTableColumn<dynamic>, double> colWidths) {
    var sortColumn = sortColumnIndex == null ? null : columns[sortColumnIndex!];

    return Row(
      children: [
        for (var column in columns)
          InkWell(
            onTap: () => trySort(column),
            child: _ColumnHeader(
              title: column.label,
              icon: column.icon,
              width: colWidths[column],
              height: widget.rowHeight,
              alignment: column.alignment,
              sortAscending: column == sortColumn ? sortAscending : null,
            ),
          ),
      ],
    );
  }

  ListView createRowsListView(
    BuildContext context,
    Map<VTableColumn<dynamic>, double> colWidths,
  ) {
    final theme = Theme.of(context);
    final rowSeparator = BoxDecoration(
      border: Border(top: BorderSide(color: theme.dividerColor)),
    );

    final colorScheme = theme.colorScheme;
    final darkMode = colorScheme.brightness == Brightness.dark;

    return ListView.builder(
      controller: scrollController,
      itemCount: sortedItems.length,
      itemExtent: widget.rowHeight,
      itemBuilder: (BuildContext context, int index) {
        T item = sortedItems[index];
        final selected = item == selectedItem.value;
        return Container(
          key: ValueKey(item),
          color: selected ? Theme.of(context).hoverColor : null,
          child: InkWell(
            onTap: () => _select(item),
            onDoubleTap: () => _doubleTap(item),
            child: DecoratedBox(
              decoration: (widget.showHeaders || index != 0)
                  ? rowSeparator
                  : const BoxDecoration(),
              child: Row(
                children: [
                  for (var column in columns)
                    Padding(
                      padding: const EdgeInsets.only(top: 1, right: 1),
                      child: SizedBox(
                        height: widget.rowHeight - 1,
                        width: colWidths[column]! - 1,
                        child: Tooltip(
                          message: column.validate(item)?.message ?? '',
                          waitDuration: widget.tooltipDelay,
                          child: Container(
                            alignment: column.alignment ?? Alignment.centerLeft,
                            padding: const EdgeInsets.symmetric(
                              horizontal: VTable._horizPadding,
                              vertical: VTable._vertPadding,
                            ),
                            color: column
                                .validate(item)
                                ?.colorForSeverity(darkMode: darkMode),
                            child: column.widgetFor(context, item),
                          ),
                        ),
                      ),
                    )
                ],
              ),
            ),
          ),
        );
      },
    );
  }

  void trySort(VTableColumn<T> column) {
    if (!column.supportsSort) {
      return;
    }

    setState(() {
      int newIndex = columns.indexOf(column);
      if (sortColumnIndex == newIndex) {
        sortAscending = !sortAscending;
      } else {
        sortAscending = true;
      }

      sortColumnIndex = newIndex;
      column.sort(sortedItems, ascending: sortAscending);
    });
  }

  Map<VTableColumn, double> _layoutColumns(BoxConstraints constraints) {
    double width = constraints.maxWidth;

    Map<VTableColumn, double> widths = {};
    double minColWidth = 0;
    double totalGrow = 0;

    for (var col in columns) {
      minColWidth += col.width;
      totalGrow += col.grow;

      widths[col] = col.width.toDouble();
    }

    width -= minColWidth;

    if (width > 0 && totalGrow > 0) {
      for (var col in columns) {
        if (col.grow > 0) {
          var inc = width * (col.grow / totalGrow);
          widths[col] = widths[col]! + inc;
        }
      }
    }

    return widths;
  }

  void _select(T item) {
    if (widget.supportsSelection) {
      setState(() {
        if (selectedItem.value != item) {
          selectedItem.value = item;
        } else {
          selectedItem.value = null;
        }
      });
    }
  }

  void _doubleTap(T item) {
    if (widget.onDoubleTap != null) {
      widget.onDoubleTap!(item);
    }
  }
}

class _ColumnHeader extends StatelessWidget {
  final String title;
  final double? width;
  final double? height;
  final IconData? icon;
  final Alignment? alignment;
  final bool? sortAscending;

  const _ColumnHeader({
    required this.title,
    required this.width,
    required this.height,
    this.icon,
    this.alignment,
    this.sortAscending,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    var swapSortIconSized = alignment != null && alignment!.x > 0;

    return SizedBox(
      width: width,
      height: height,
      child: Container(
        padding: const EdgeInsets.symmetric(
          horizontal: VTable._horizPadding,
          vertical: VTable._vertPadding,
        ),
        //alignment: alignment ?? Alignment.centerLeft,
        child: Row(
          //mainAxisSize: MainAxisSize.min,
          children: [
            if (sortAscending != null && swapSortIconSized)
              AnimatedRotation(
                turns: sortAscending! ? 0 : 0.5,
                duration: const Duration(milliseconds: 200),
                child: const Icon(Icons.keyboard_arrow_up),
              ),
            Expanded(
              child: icon != null
                  ? Tooltip(
                      message: title,
                      child: Align(
                        alignment: alignment ?? Alignment.centerLeft,
                        child: Icon(icon),
                      ),
                    )
                  : Text(
                      title,
                      style: const TextStyle(fontWeight: FontWeight.bold),
                      textAlign: swapSortIconSized ? TextAlign.end : null,
                    ),
            ),
            if (sortAscending != null && !swapSortIconSized)
              AnimatedRotation(
                turns: sortAscending! ? 0 : 0.5,
                duration: const Duration(milliseconds: 200),
                child: const Icon(Icons.keyboard_arrow_up),
              ),
          ],
        ),
      ),
    );
  }
}

/// A callback to render a data object (`T`) to a widget.
typedef RenderFunction<T> = Widget? Function(
    BuildContext context, T object, String out);

/// A callback to render a data object to a string.
typedef TransformFunction<T> = String Function(T object);

/// A callback to apply style to a cell based on a data object's properties.
typedef StyleFunction<T> = TextStyle? Function(T object);

/// A callback to compare two table row data objects.
typedef CompareFunction<T> = int Function(T a, T b);

/// A callback to apply validation to a table cell.
///
/// See [ValidationResult] for the possible return values.
typedef ValidationFunction<T> = ValidationResult? Function(T object);

/// The definition of a table's column.
class VTableColumn<T> {
  /// The title of a column.
  final String label;

  /// An icon for the column; if provided, this will be displayed instead of the
  /// column's title.
  final IconData? icon;

  /// The desired width of a column in pixels.
  final int width;

  /// The grow of a column; this is used to allocate excess table width between
  /// columns.
  final double grow;

  /// The alignment for a column's contents; this will default to
  /// [Alignment.centerLeft].
  final Alignment? alignment;

  /// A callback to convert a cell's contents to a string; if not provided, this
  /// will default to calling `toString` on the row object.
  final TransformFunction<T>? transformFunction;

  /// A callback to allow for custom styling of a cell (bold, italic, ...).
  final StyleFunction<T>? styleFunction;

  /// A callback to compare two row objects; this is used when sorting rows.
  final CompareFunction<T>? compareFunction;

  /// A set of validators for cell contents.
  final List<ValidationFunction<T>> validators;

  /// A callback to support converting a data object to a custom widget.
  final RenderFunction<T>? renderFunction;

  /// Construct a new [VTableColumn intance.
  VTableColumn({
    required this.label,
    required this.width,
    this.icon,
    this.alignment,
    this.grow = 0,
    this.transformFunction,
    this.styleFunction,
    this.compareFunction,
    this.validators = const [],
    this.renderFunction,
  });

  Widget widgetFor(BuildContext context, T item) {
    final out = transformFunction != null ? transformFunction!(item) : '$item';

    if (renderFunction != null) {
      Widget? widget = renderFunction!(context, item, out);
      if (widget != null) return widget;
    }

    var style = styleFunction == null ? null : styleFunction!(item);
    return Text(
      out,
      style: style,
      maxLines: 2,
    );
  }

  void sort(List<T> items, {required bool ascending}) {
    if (compareFunction != null) {
      items
          .sort(ascending ? compareFunction : (a, b) => compareFunction!(b, a));
    } else if (transformFunction != null) {
      items.sort((T a, T b) {
        var strA = transformFunction!(a);
        var strB = transformFunction!(b);
        return ascending ? strA.compareTo(strB) : strB.compareTo(strA);
      });
    }
  }

  bool get supportsSort => compareFunction != null || transformFunction != null;

  ValidationResult? validate(T item) {
    if (validators.isEmpty) {
      return null;
    } else if (validators.length == 1) {
      return validators.first(item);
    } else {
      List<ValidationResult> results = [];
      for (var validator in validators) {
        ValidationResult? result = validator(item);
        if (result != null) {
          results.add(result);
        }
      }
      return ValidationResult.combine(results);
    }
  }
}

/// The possible severity results for cell validation.
enum Severity {
  info,
  note,
  warning,
  error,
}

/// A cell validation result; this includes both the severity and a custom
/// message.
///
/// E.g., `ValidationResult('gravity too high', Severity.warning)`.
class ValidationResult {
  final String message;
  final Severity severity;

  ValidationResult(this.message, this.severity);

  factory ValidationResult.error(String message) =>
      ValidationResult(message, Severity.error);

  factory ValidationResult.warning(String message) =>
      ValidationResult(message, Severity.warning);

  factory ValidationResult.note(String message) =>
      ValidationResult(message, Severity.note);

  factory ValidationResult.info(String message) =>
      ValidationResult(message, Severity.info);

  Color colorForSeverity({bool darkMode = true}) {
    if (darkMode) {
      switch (severity) {
        case Severity.info:
          return Colors.grey.shade400;
        case Severity.note:
          return Colors.blue.shade400;
        case Severity.warning:
          return Colors.yellow.shade700;
        case Severity.error:
          return Colors.red.shade400;
      }
    } else {
      switch (severity) {
        case Severity.info:
          return Colors.grey.shade400.withAlpha(127);
        case Severity.note:
          return Colors.blue.shade200.withAlpha(127);
        case Severity.warning:
          return Colors.yellow.shade200.withAlpha(127);
        case Severity.error:
          return Colors.red.shade300.withAlpha(127);
      }
    }
  }

  @override
  String toString() => '$severity $message';

  static ValidationResult? combine(List<ValidationResult> results) {
    if (results.isEmpty) {
      return null;
    } else if (results.length == 1) {
      return results.first;
    } else {
      String message = results.map((r) => r.message).join('\n');
      Severity severity = results
          .map((r) => r.severity)
          .reduce((a, b) => a.index >= b.index ? a : b);
      return ValidationResult(message, severity);
    }
  }
}
